src/main/java/ch/njol/skript/expressions/ExprTool.java

```diff
@@ -42,75 +30,120 @@
 		setExpr((Expression<LivingEntity>) exprs[0]);
-		offHand = parser.mark == 1;
+		offHand = parseResult.hasTag("offhand");
 		return true;
 	}
 
 	@Override
-	protected Slot[] get(final Event e, final LivingEntity[] source) {
-		final boolean delayed = Delay.isDelayed(e);
+	protected Slot[] get(Event event, LivingEntity[] source) {
+		boolean delayed = Delay.isDelayed(event);
 		return get(source, new Getter<Slot, LivingEntity>() {
 			@Override
 			@Nullable
-			public Slot get(final LivingEntity ent) {
+			public Slot get(LivingEntity entity) {
 				if (!delayed) {
-					if (!offHand && e instanceof PlayerItemHeldEvent && ((PlayerItemHeldEvent) e).getPlayer() == ent) {
-						final PlayerInventory i = ((PlayerItemHeldEvent) e).getPlayer().getInventory();
-						return new InventorySlot(i, getTime() >= 0 ? ((PlayerItemHeldEvent) e).getNewSlot() : ((PlayerItemHeldEvent) e).getPreviousSlot());
-					} else if (e instanceof PlayerBucketEvent && ((PlayerBucketEvent) e).getPlayer() == ent) {
-						final PlayerInventory i = ((PlayerBucketEvent) e).getPlayer().getInventory();
-						boolean isOffHand = ((PlayerBucketEvent) e).getHand() == org.bukkit.inventory.EquipmentSlot.OFF_HAND || offHand;
-						return new InventorySlot(i, isOffHand ? EquipmentSlot.EquipSlot.OFF_HAND.slotNumber
-							: ((PlayerBucketEvent) e).getPlayer().getInventory().getHeldItemSlot()) {
+					if (offHand && event instanceof InventoryClickEvent && ((InventoryClickEvent) event).getWhoClicked().equals(entity) && getTime() == 1) {
+						InventoryClickEvent inventoryClickEvent = (InventoryClickEvent) event;
+						// When a player uses a number key to swap an item from hotbar to offhand. This simplifies the process with future states.
+						if (inventoryClickEvent.getClick() == ClickType.NUMBER_KEY && inventoryClickEvent.getSlot() == EquipmentSlot.EquipSlot.OFF_HAND.slotNumber) {
+							PlayerInventory inventory = inventoryClickEvent.getWhoClicked().getInventory();
+							return new InventorySlot(inventory, inventoryClickEvent.getHotbarButton());
+						} else if (inventoryClickEvent.getClick() == ClickType.SWAP_OFFHAND) {
+							PlayerInventory inventory = inventoryClickEvent.getWhoClicked().getInventory();
+							return new InventorySlot(inventory, inventoryClickEvent.getSlot());
+						} else if (inventoryClickEvent.getSlot() == EquipmentSlot.EquipSlot.OFF_HAND.slotNumber) {
+							switch (inventoryClickEvent.getAction()) {
+								case NOTHING: // When you double click to collect to cursor, it's not COLLECT_TO_CURSOR...
+								case PICKUP_ALL:
+								case PICKUP_HALF:
+								case PICKUP_ONE:
+								case PICKUP_SOME:
+									return new InventorySlot(inventoryClickEvent.getClickedInventory(), inventoryClickEvent.getRawSlot());
+								case PLACE_ALL:
+								case PLACE_ONE:
+								case PLACE_SOME:
+								case SWAP_WITH_CURSOR:
+									return new CursorSlot((Player) inventoryClickEvent.getWhoClicked());
```

Comment:
This isn't accurate in most cases. It only works if:
a) it's swap with cursor
b) it's place all and the slot was previously empty
c) it's place one and the cursor only had one item and the slot was empty.

Basically it's not taking into account the existing items in the offhand and how many will actually be transferred.